Truncate page hit title

kuzmany opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 4.x)
* a.b for any bug fixes (e.g. 4.0, 4.1, 4.2)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

If you use page with tilte longer like 191 chars, then tracking pixel is not triggered. This PR fixed it

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create page with title longer like 191 chars like this

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars Title longer than 191 chars </title>

</head>
<body>

</html>
```
3. Put tracking pixel on the page
4. Open the page
5. Tracking pixel is not triggered
6. After PR tracking pixel is triggered with title of page truncated to 191 chars

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10709"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

kathrin-schmid committed:
Update language string (#10684)

The button that i updated is for saving changes made in the "Edit code" window and should thus say "Save", not "Edit".

escopecz committed:
Fixes 500 error in forms when render style is deactivated against 4.1 (#10697)

* Fixes issue #10453

* Adding functional test covering the bug

https://github.com/mautic/mautic/issues/10453

Co-authored-by: Rolando <rpayanm@gmail.com>

escopecz committed:
Fixing issues when using PHPSTAN with more processes (#10696)

escopecz committed:
Bumping to version 4.1.1

RCheesley committed:
Remove reference to mautibox (#10681)

kuzmany committed:
Trucnate page hit title

kuzmany committed:
Add unit tests

kuzmany committed:
Fix PHPStan

mmarton commented:
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

- mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?) 
> 500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1 

- Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

- this is my first review, please correct me if I made any mistakes. Thankyou

mmarton commented:
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

    mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?)

    500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1

    Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

    this is my first review, please correct me if I made any mistakes. Thankyou

mmarton reviewed (commented):
Hi!

I've tested this with title lengths 190 and 2000 characters (linux mint, firefox 95, chrome 96)

I could not confirm the original issue. The tracking worked for me with any title length.

sidenotes:

 - mautic throws an exception if you try to save a page with title longer than 190 (could use a length validation, is this PR a good candidate for that or should I open a new issue?)
> 500 Internal Server Error - An exception occurred while executing 'UPDATE pages SET date_modified = ?, checked_out = ?, checked_out_by = ?, title = ?, revision = ? WHERE id = ?' with params ["2021-12-29 14:23:35", null, null, "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012", 4, 1]: SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'title' at row 1

- Could't use gitpod for testing because the pagebuilder throw a cors js error, so I tested locally, is there a way to make it work?

- this is my first review, please correct me if I made any mistakes. Thankyou

tejasync committed:
Fix lock wait deleting unused ips (#10710)

* Deleted unused IPs in batches and some improvements.

* Added return type declaration to command methods and some improvements.

* Removed dump()

* Created separate methods for get and delete unused ips.

* Consumed repo methods in model.

* Updated command to use model methods.

* Added test case.

* Fix static analysis

escopecz commented on app/bundles/PageBundle/Entity/Hit.php:
```diff
@@ -558,6 +558,7 @@ public function getUrl()
      */
     public function setUrlTitle($urlTitle)
     {
+        $urlTitle       = mb_strlen($urlTitle) <= 191 ? $urlTitle : substr($urlTitle, 0, 191);
```
Let's use the multibyte functions
```suggestion
        $urlTitle       = mb_strlen($urlTitle) <= 191 ? $urlTitle : mb_substr($urlTitle, 0, 191);
```

escopecz commented on app/bundles/PageBundle/Tests/Entity/HitTest.php:
```diff
@@ -0,0 +1,31 @@
+        $hit = new Hit();
+        $hit->setUrlTitle($urlTitle);
+
+        Assert::assertEquals($expected, strlen($hit->getUrlTitle()));
```
```suggestion
        Assert::assertEquals($expected, mb_substr($hit->getUrlTitle()));
```

escopecz reviewed (changes requested):
Thanks Zdeno! 2 small suggestions.

kuzmany committed:
Update app/bundles/PageBundle/Entity/Hit.php

Co-authored-by: John Linhart <jan.linhart@acquia.com>

kuzmany committed:
Update app/bundles/PageBundle/Tests/Entity/HitTest.php

Co-authored-by: John Linhart <jan.linhart@acquia.com>

escopecz commented:
@mmarton can you please check that you were able to store a page title longer than 255 characters in the page_hits table before this change? I found an issue (https://github.com/mautic/mautic/issues/7367) reporting the same problem.

And yes, the error you found should be fixed in another PR. It's not directly related to this one.

escopecz reviewed (dismissed):
The code make sense, I can see the `page_hits.url_title` column is type of `VARCHAR(191)`, the test confirms it works as expected 👍

escopecz commented on app/bundles/PageBundle/Tests/Entity/HitTest.php:
```diff
@@ -0,0 +1,31 @@
+        $hit = new Hit();
+        $hit->setUrlTitle($urlTitle);
+
+        Assert::assertEquals($expected, mb_substr($hit->getUrlTitle()));
```
OOps, I meant
```suggestion
        Assert::assertEquals($expected, mb_strlen($hit->getUrlTitle()));
```

kuzmany committed:
Update app/bundles/PageBundle/Tests/Entity/HitTest.php

Co-authored-by: John Linhart <jan.linhart@acquia.com>

escopecz reviewed (dismissed):
Re-approving

rahuld-dev committed:
Fix contact batch api for single id value (#10700)

Co-authored-by: John Linhart <admin@escope.cz>

abailey-dev committed:
fix(LeadTimelineEvent): keys do not match the func args, strips keys (#10663)

Fix error when using PHP 8:
{"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): array_merge() does not accept unknown named parameters}

Not sure if this is the most elegant way to address the issue but this bug has broken the site functionality.

tejasync committed:
User update role api (#10668)

* Allow user to update role via API

* Update user's role using API: test cases

* Update user's role using API: added more test cases

* Fix test cases and add more test case

* Added AbstractMauticTestCase::loginUser() method that allows login user in non-restricted areas

* Fix failing tests

Co-authored-by: fedys <miroslav.fedeles@gmail.com>
Co-authored-by: John Linhart <admin@escope.cz>

kuzmany committed:
Fix send unpublished SMS (#10577)

* Fix send unpublished SMS

* Fix send unpublished SMS for broadcast command

* Add CampaignSendSubscriberTest

* Add testSendSmsNotPublished test

* Minor change to tests

* Fixing types

* Automatic cs fix

Co-authored-by: John Linhart <admin@escope.cz>

kuzmany committed:
Fix group by If you use count columns for assets download (#10693)

* Fix group by If you use count columns

* Add unit tests

* Fix unit tests

* Remove define from tests

* Fix unit tests

* Fix unit tests

* Fix unit tests

Co-authored-by: John Linhart <admin@escope.cz>

eloimarquessilva committed:
Update grapesjs-custom.css (#10728)

Removing the background color f0f0f0 from gjs-field-radio class to improve the color constrast between radio's background and radio's options and returning the class parameters to the default of the original editor.

Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

RCheesley committed:
Update stale.yml (#10746)

Apply Stalebot only to issues and not PRs

Co-authored-by: John Linhart <admin@escope.cz>

mollux committed:
Fixed variable name (#10753)

Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>
Co-authored-by: Joey Keller <jos0405@gmail.com>

escopecz committed:
Change field definition for column_value to longtext instead of varchar (#10778)

* Change field definition for column_value to longtext instead of varchar 255

* Update app/migrations/Version20200729170800.php

Co-authored-by: John Linhart <admin@escope.cz>

* I suggested bad namespace path during code review. Fixing it here

Co-authored-by: Don Gilbert <don@dongilbert.net>

tomekkowalczyk committed:
Correct bug no forms in the select contact source campaign when the forms have identical names (#10717)

* Correct name form Campaign Contact Source

* Refactoring and add test

* Finishing up the test

* Fixing the bug discovered by the test

* CS fix

Co-authored-by: John Linhart <admin@escope.cz>
Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

patrykgruszka committed:
Restore X value in contact boolean fields (#10716)

* Fix empty value in contact boolean fields

* Fix configuration page save

* Fix controller functional tests

* Always retype the isEnabled param to int. Even for PATCH

* Fixing phpstan issue

  Line   app/bundles/ChannelBundle/Controller/Api/MessageApiController.php
 ------ -------------------------------------------------------------------
  32     Property
         Mautic\ChannelBundle\Controller\Api\MessageApiController::$model
         (Mautic\ChannelBundle\Model\MessageModel) does not accept
         Mautic\CoreBundle\Model\AbstractCommonModel.

* Fixing PHP Notice - Undefined index: email

"file":"\/var\/www\/html\/mautic\/app\/bundles\/ChannelBundle\/Controller\/Api\/MessageApiController.php","line":54

I cannot figure out why the condition for PATCH was there in the first place. It's causing only troubles.

* Adding functional API tests covering the problematic endpoint and method, fixing the controller

* Fix csfixer issue

* Adding missing description to payloads for assertion

Co-authored-by: John Linhart <admin@escope.cz>
Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

RCheesley committed:
Add missing information (#10656)

* Add missing information

* Missing commas :blush:

TonyBogdanov committed:
Fix for pushing to integrations in campaign actions (#10674)

* Fix for #10517

Some integrations such as Hubspot have an `'encode_parameters' => 'json'` setting, which triggers the parameters being sent in `POST` / `PUT` / `PATCH` requests to be `json_encode`-d: https://github.com/mautic/mautic/blob/5975231722cfde8bdad783b373c70ad20333db37/app/bundles/PluginBundle/Integration/AbstractIntegration.php#L760

That causes a problem when the result is passed to Guzzle's `request` method under key `\GuzzleHttp\RequestOptions::FORM_PARAMS`, because the payload is expected to be an array: https://github.com/mautic/mautic/blob/5975231722cfde8bdad783b373c70ad20333db37/app/bundles/PluginBundle/Integration/AbstractIntegration.php#L829 as per the documentation: https://docs.guzzlephp.org/en/stable/request-options.html#form-params

The above RP addresses this by passing the payload under `\GuzzleHttp\RequestOptions::BODY` instead, **if** the payload is a string, rather than an array.

Tested and working with HubSpot.

* Adding test to the payload key fix

* Creating Client in another method so it could be mocked

Plus some code style improvements

* CS Fixer

* STAN fixes

* Removing whitespace

Co-authored-by: John Linhart <admin@escope.cz>
Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

RCheesley committed:
Bump to 4.1.2

RCheesley committed:
Merge branch '4.1' into truncate-page-hit-title

RCheesley committed:
Merge branch '4.x' into truncate-page-hit-title

mabumusa1 committed:
Merge branch '4.x' into truncate-page-hit-title

mabumusa1 reviewed (approved):
I did a code review, I think it works and solve a problem for long URL titles

kuzmany committed:
Merge branch '4.x' into truncate-page-hit-title

kuzmany committed:
Merge branch '4.x' into truncate-page-hit-title
